src/Xrefcheck/Core.hs

```diff
@@ -198,51 +189,73 @@ data DirectoryStatus
     -- ^ Directories containing those files.
-  , riRoot :: CanonicalPath
-    -- ^ Repository root.
   }
 
 -- Search for a file in the repository.
-lookupFile :: CanonicalPath -> RepoInfo -> Maybe FileStatus
+lookupFile :: RelPosixLink -> RepoInfo -> Maybe FileStatus
 lookupFile path RepoInfo{..} =
   M.lookup path riFiles
 
 -- Search for a directory in the repository.
-lookupDirectory :: CanonicalPath -> RepoInfo -> Maybe DirectoryStatus
+lookupDirectory :: RelPosixLink -> RepoInfo -> Maybe DirectoryStatus
 lookupDirectory path RepoInfo{..} =
   M.lookup path riDirectories
 
 -----------------------------------------------------------
 -- Instances
 -----------------------------------------------------------
 
-instance NFData Position
-instance NFData Reference
 instance NFData ReferenceInfo
-instance NFData AnchorType
 instance NFData Anchor
+instance NFData AnchorType
+instance NFData ExternalLink
 instance NFData FileInfo
+instance NFData FileLink
+instance NFData Position
+instance NFData Reference
+instance NFData ReferenceInfoFile
 
 instance Given ColorMode => Buildable Reference where
   build Reference{..} =
-    [int||
-    reference #{paren . build $ rInfo} #{rPos}:
-      - text: #s{rName}
-      - link: #{if null rLink then "-" else rLink}
-      - anchor: #{rAnchor ?: styleIfNeeded Faint "-"}
-    |]
-
-instance Given ColorMode => Buildable ReferenceInfo where
-  build = \case
-    RIFileLocal -> colorIfNeeded Green "file-local"
-    RIFileRelative -> colorIfNeeded Yellow "relative"
-    RIFileAbsolute -> colorIfNeeded Blue "absolute"
-    RIExternal -> colorIfNeeded Red "external"
-    RIOtherProtocol -> ""
+    case rInfo of
+      RIFile ReferenceInfoFile{..} ->
+        case rifLink of
+          FLLocal ->
+            [int||
+            reference #{paren $ colorIfNeeded Green "file-local"} #{rPos}:
+              - text: #s{rName}
+              - anchor: #{rifAnchor ?: styleIfNeeded Faint "-"}
+            |]
+          FLRelative link ->
+            [int||
+            reference #{paren $ colorIfNeeded Yellow "relative"} #{rPos}:
+              - text: #s{rName}
+              - link: #{if null link then "-" else link}
+              - anchor: #{rifAnchor ?: styleIfNeeded Faint "-"}
+            |]
+          FLAbsolute link ->
+            [int||
+            reference #{paren $ colorIfNeeded Yellow "absolute"} #{rPos}:
+              - text: #s{rName}
+              - link: #{if null link then "-" else "/" <> link}
+              - anchor: #{rifAnchor ?: styleIfNeeded Faint "-"}
+            |]
+      RIExternal (ELUrl url) ->
+        [int||
+        reference #{paren $ colorIfNeeded Red "external"} #{rPos}:
+          - text: #s{rName}
+          - link: #{if null url then "-" else url}
+        |]
+      RIExternal (ELOther url) ->
+        [int||
+        reference (other) #{rPos}:
+          - text: #s{rName}
+          - link: #{if null url then "-" else url}
+        |]
 
```

Comment:
Here I have removed unnecessary fields from reference types. For example:
- Link in local references.
- Anchor in external references. It is now part of the URL.

Do you agree with this output change?

Reply:
That's really interesting.

Strictly speaking, external links also can have anchors, but we do not account for them at all, so not displaying them seems reasonable. I would say, it is better than before.

And for link in local references - I also agree.

Nice that parsing the link into `ReferenceInfo` let us for such fine-tuned pretty-printing and stuff.

Reply:
It is also a win if we go on with the symlink scanner (#262). With this changes, it would be easier to model and display them.